NEXT-20683 - Add fastly docs (#547)
shopware/docs: guides/hosting/infrastructure/reverse-http-cache.md

#set resp.http.X-Cache-Hits = obj.hits;
```

## Configure Fastly

Fastly is supported since Shopware 6.4.11.0 out of the box with some configurations. To enable it we need to create a new file in `config/packages/storefront.yaml`

```yaml
storefront:
    csrf:
        enabled: true
        mode: ajax
    reverse_proxy:
        enabled: true
        fastly:
          enabled: true
          api_key: '<personal-token-from-fastly>'
          service_id: '<service-id>'
```

### Fastly VCL Snippets

Additionally we need to setup some VCL Snippets in the Fastly interface:

**Name:** Normalize URLs

**Subroutine:** vcl_recv

```
# Mitigate httpoxy application vulnerability, see: https://httpoxy.org/
unset req.http.Proxy;

# Strip query strings only needed by browser javascript. Customize to used tags.

> if (req.url ~ "(\?|&)(pk_campaign|piwik_campaign|pk_kwd|piwik_kwd|pk_keyword|pixelId|kwid|kw|adid|chl|dv|nk|pa|camid|adgid|cx|ie|cof|siteurl|utm_[a-z]+|_ga|gclid)=") {
>     # see rfc3986#section-2.3 "Unreserved Characters" for regex
>     set req.url = regsuball(req.url, "(pk_campaign|piwik_campaign|pk_kwd|piwik_kwd|pk_keyword|pixelId|kwid|kw|adid|chl|dv|nk|pa|camid|adgid|cx|ie|cof|siteurl|utm_[a-z]+|_ga|gclid)=[A-Za-z0-9\-\_\.\~]+&?", "");
> }
> set req.url = regsub(req.url, "(\?|\?&|&)$", "");

Reviewer: While technically correct, this is a lot more verbose and heavy handed (use of 2 regexes) approach than necessary. You can simply use [querystring.regfilter](https://developer.fastly.com/reference/vcl/functions/query-string/querystring-regfilter/) to accomplish this.

```suggestion
if (req.url != req.url.path) {
  set req.url = querystring.regfilter(req.url, "pk_campaign|piwik_campaign|pk_kwd|piwik_kwd|pk_keyword|pixelId|kwid|kw|adid|chl|dv|nk|pa|camid|adgid|cx|ie|cof|siteurl|utm_[a-z]+|_ga|gclid");
}
```

# Normalize query arguments
set req.url = querystring.sort(req.url);

# Make sure that the client ip is forward to the client.
if (req.http.x-forwarded-for) {
    set req.http.X-Forwarded-For = req.http.X-Forwarded-For + ", " + client.ip;
} else {
    set req.http.X-Forwarded-For = client.ip;
}

# Normalize Accept-Encoding header
# straight from the manual: https://www.varnish-cache.org/docs/3.0/tutorial/vary.html
if (req.http.Accept-Encoding) {
    if (req.url ~ "\.(jpg|png|gif|gz|tgz|bz2|tbz|mp3|ogg)$") {
        # No point in compressing these
        unset req.http.Accept-Encoding;
    } elsif (req.http.Accept-Encoding ~ "gzip") {
        set req.http.Accept-Encoding = "gzip";
    } elsif (req.http.Accept-Encoding ~ "deflate") {
        set req.http.Accept-Encoding = "deflate";
    } else {
        # unkown algorithm
        unset req.http.Accept-Encoding;
    }

> }

Reviewer: These are not necessary for Fastly. [We are normalizing `Accept-Encoding`](https://developer.fastly.com/reference/http/http-headers/Accept-Encoding/) and [we are not compressing those extensions by default](https://docs.fastly.com/en/guides/enabling-automatic-gzipping). If you still want to force `Accept-Encoding` removal for listed extensions, please use `req.url.ext`, thus something like this:

```
    if (req.url.ext ~ "^(jpg|png|gif|gz|tgz|bz2|tbz|mp3|ogg)$") {
        # No point in compressing these
        unset req.http.Accept-Encoding;
    }
```
However, I don't feel this is necessary.

> if (req.method != "GET" &&
>     req.method != "HEAD" &&
>     req.method != "PUT" &&
>     req.method != "POST" &&
>     req.method != "TRACE" &&
>     req.method != "OPTIONS" &&
>     req.method != "PATCH" &&
>     req.method != "DELETE") {
>     /* Non-RFC2616 or CONNECT which is weird. */
>     return (pass);
> }
> 
> # We only deal with GET and HEAD by default
> if (req.method != "GET" && req.method != "HEAD") {
>     return (pass);
> }

Reviewer: The starting point for Fastly services, [called boilerplate](https://developer.fastly.com/learning/vcl/using/#adding-vcl-to-your-service-configuration), looks like this 

```
sub vcl_recv {
#FASTLY recv

  # Normally, you should consider requests other than GET and HEAD to be uncacheable
  # (to this we add the special FASTLYPURGE method)
  if (req.method != "HEAD" && req.method != "GET" && req.method != "FASTLYPURGE") {
    return(pass);
  }
```
You can notice similar code. The snippet will be placed into the expanded `#FASTLY recv` macro, thus prior to the code. For this reason your addition might be useful, as you're doing your own `return(lookup)` and not letting it to fall through. However I'd recommend having the same code we have instead. Beside the consistency, the main reason is to allow for `FASTLYPURGE`; this is where a `PURGE` method is mapped to. That will allow a single object instant purge to work. 
```suggestion
  # Normally, you should consider requests other than GET and HEAD to be uncacheable
  # (to this we add the special FASTLYPURGE method)
  if (req.method != "HEAD" && req.method != "GET" && req.method != "FASTLYPURGE") {
    return(pass);
  }
```

# Don't cache Authenticate & Authorization
if (req.http.Authenticate || req.http.Authorization) {
    return (pass);
}

# Always pass these paths directly to php without caching
# Note: virtual URLs might bypass this rule (e.g. /en/checkout)

> if (req.url ~ "^/(checkout|account|admin|api)(/.*)?$") {

Reviewer: You seems to be evaluating URL path here, use [req.url.path](https://developer.fastly.com/reference/vcl/variables/client-request/req-url-path/) instead
```suggestion
if (req.url.path ~ "^/(checkout|account|admin|api)(/.*)?$") {
```

    return (pass);
}

return (lookup);
```

**Name:** Add Shopware Custom Hashing

**Subroutine:** vcl_hash

```
# Consider Shopware http cache cookies

> if (req.http.cookie ~ "sw-cache-hash=") {
>     set req.hash += regsub(req.http.cookie, "^.*?sw-cache-hash=([^;]*);*.*$", "\1");
> } elseif (req.http.cookie ~ "sw-currency=") {
>     set req.hash += regsub(req.http.cookie, "^.*?sw-currency=([^;]*);*.*$", "\1");
> }

Reviewer: ```suggestion
if (req.http.cookie:sw-cache-hash) {
  set req.hash += req.http.cookie:sw-cache-hash;
} elseif (req.http.cookie:sw-currency) {
  set req.hash += req.http.cookie:sw-currency;
}
```
In Fastly VCL we have [subfield](https://developer.fastly.com/reference/vcl/functions/miscellaneous/subfield/). No need to run regexes twice.

```

**Name:** Respect Shopware States to Pass the cache

**Subroutine:** vcl_hit

```

> if (req.http.cookie ~ "sw-states=") {
>    set req.http.states = regsub(req.http.cookie, "^.*?sw-states=([^;]*);*.*$", "\1");

Reviewer: Same as the previous comment

```suggestion
if (req.http.cookie:sw-states) {
   set req.http.states = req.http.cookie:sw-states;
```

>    if (req.http.states ~ "logged-in" && obj.http.sw-invalidation-states ~ "logged-in" ) {

Reviewer: ```suggestion
   if (std.strstr(req.http.states,"logged-in") &&  std.strstr(obj.http.sw-invalidation-states,"logged-in" ) {
```

      return (pass);
   }

>    if (req.http.states ~ "cart-filled" && obj.http.sw-invalidation-states ~ "cart-filled" ) {

Reviewer: ```suggestion
   if (std.str(req.http.states,"cart-filled") && std.strstr(obj.http.sw-invalidation-states,"cart-filled" )) {
```
